RFC: rewrite on golang. extend functionality and useability.

AntonioMeireles opened the pull request:
work in progress, quick hack done while babysitting, ... releasing early after chatting with @ecnahc515 on IRC.

AntonioMeireles commented:
@philips @bakins @ecnahc515
atached patchset _may_ be starting to look sane and reasonable enough to warrant the begginings of a review. Thanks in advance.

philips commented on commands.go:
```diff
@@ -0,0 +1,77 @@
+
+package main
+
+import "github.com/codegangsta/cli"
```
would you mind using github.com/spf13/cobra instead? We have started to use it on all of our projects.

AntonioMeireles replied:
sure, will look at it, not religious about codegangsta/cli. (fwiw, i went to it because i used in the past and etcd was also using it)

AntonioMeireles replied:
consuming already spf13/{viper,cobra} in latest iteration.

philips commented:
This is looking really cool, I will try and review it more in-depth soon.

chancez commented on cloud-init/docker-only-with-persistent-storage.txt:
```diff
@@ -0,0 +1,79 @@
+        [Install]
+        WantedBy=sockets.target
+
+    - name: format-ephemeral.service
```
This and `persistent-vda.mount` can be replaced by mounting in a disk and passing `root=/dev/vda` or `root=LABEL=<disk-label>` in the kernel cmdline. The PXE image doesn't have a rootfs, and supports providing your own root disk.

AntonioMeireles replied:
i know ... fwiw i found that sub optimal as too wide reaching. i wanted the ability to take full advantage of the simplicity of PXE boot (as each time _everything_ will work same way, no leftovers, etc) while having the ability to store persistently _just_ what i wanted with _a la carte_ granularity. (say ones want 1 vol for docker, another to  rkt, etc, to test version migrations/whatever). anyway will look onto extending current scheme to cover all angles. :-)

chancez commented on helpers.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+// returns true if payload not nil/empty
+func got(in interface{}) bool {
```
I'm not really a huge fan of this function. It isn't all that idiomatic, it's usually just best to check what you want specifically at the caller location.

AntonioMeireles replied:
matter of habits. i usually reuse it on my hacks, as it spares a few bits. ~~will get~~ got rid of it already in latest push.

AntonioMeireles committed:
rewrite on golang. extend functionality and useability.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
rewrite on golang. extend functionality and useability.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles commented:
@philips @ecnahc515 when you have time and/or opportunity...
@rimusz the json hooks there enough for your UI stuff ?

thanks in advance.

AntonioMeireles committed:
avert panic on rm when there's nothing to delete

AntonioMeireles committed:
avert panic on rm when there's nothing to delete

AntonioMeireles committed:
harden version handling

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
harden version handling

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
sanitize VM pid recording

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
sanitize VM pid recording

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor images' processing

native, without external deps, now.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor images' processing

native, without external deps, now.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
clean staled data from non-running instances

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
clean staled data from non-running instances

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
ensure tap device atomicity

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
ensure tap device atomicity

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
add a bit more consistency to UI

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
add a bit more consistency to UI

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
catch all potential errors in sshKeyGen

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
catch all potential errors in sshKeyGen

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor OEMfication. fully native now too.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor OEMfication. fully native now too.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor ssh machinery. native now also.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor ssh machinery. native now also.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
simplify greatly guest's IP and Pid capture.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
simplify greatly guest's IP and Pid capture.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
update cobra, for cli's git like behaviour on mispellings.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
update cobra, for cli's git like behaviour on mispellings.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
keep terminal state sane in case attached sessions go wrong

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
keep terminal state sane in case attached sessions go wrong

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge branch 'golang' of github.com:AntonioMeireles/coreos-xhyve into golang

AntonioMeireles committed:
fix Makefile logic for deterministic doc generation

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
more UI tweaks

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
strengthen http error handling

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor queing machinery

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
more refactoring/simplification of the OEMification process

trying to behave more closely as the big boys

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
godeps' update

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
consumes now bundled, golangfied xhyve

no more external deps of any kind, so that simply app signing this will allow
full privileges escalation without sudo soonish.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
godeps' update

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

rimusz commented:
@AntonioMeireles yes `json hooks` looks good sorry for the delay on coming back.
As I started to work at Deis,  hope I will have more time for my UI App now :)
VBox and Vagrant are so much slower to start comparing with xhyve :)

AntonioMeireles committed:
godeps' update

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
adapt to latest hooklyft/xhyve upstream developments

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
remove dead code, plus cosmetics

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

pires commented:
Since Vagrant is not providing a provider for xhyve this seems to be the best approach available. But due to the size of the PR how realistic if for us to expect it to be merged @philips?

@AntonioMeireles could you remove vendored dependencies to make it easier to review? I think that anyone that wants to use this should be able to `restore` said dependencies.

AntonioMeireles commented:
@pires offering  it to get reviewed/merged here is on paper the right thing to do as this was the original upstream. 

FWIW **if** things _here_  keep evolving at the pace they had been **and** **if** there is enough interest elsewhere **then** i'm open to my fork becoming _the_ active upstream on this, etc, etc, etc.

AntonioMeireles committed:
godeps' update

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
some more cosmetic tweaks

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
don't require sudo to halt VMs.

it's only required if we need for some reason to do an hard
kill (in the cases we just lost communication with the VM, or
it just hanged)

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
godeps' update

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
fix volume use example in README to match shipping code

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
fix spelling typo (spotted by @rimusz)

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
add the ability to not download a (new) `latest` image on run calls

- per @rimusz request for his https://github.com/rimusz/coreos-osx

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
reflect ongoing rebrand everywhere.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #1 from AntonioMeireles/golang

reflect ongoing rebrand everywhere

AntonioMeireles committed:
fix Godeps/Godeps.json

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
actually reflect blob naming change when looking for running VMs

closes #2

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #3 from AntonioMeireles/golang

actually reflect blob naming change when looking for running VMs

AntonioMeireles committed:
Add an acknowledgements section to the README

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
add ability of not check for new online images to `load` payloads too

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #6 from AntonioMeireles/golang

add ability of not check for new online images to `load` payloads too

AntonioMeireles committed:
ensure deterministic boot order when using `load`

- see #5 for context.
- updated profiles/demo.toml

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #7 from AntonioMeireles/golang

ensure deterministic boot order when using `load`

AntonioMeireles committed:
dont't free stuff before actually consuming it, as it avoids a few crashes

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
on halt really wait for the VM to be deader than dead.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
workaround crash inducing weirdness in (either/or) osx/golang's net stack

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #8 from AntonioMeireles/crasher_fixes

crasher fixes

AntonioMeireles committed:
don't let cruft on the local fs after an halt

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
recover gracefully from crash when requesting a free port

- happens randomly, shouldn't, and this is supposed to get around it

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
also store MacAdress in VMInfo

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor and make more resilient UUIDs handling

in particular we workaround an odd bug where with some UUIDs we aren't
able to guess the MAC address.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #9 from AntonioMeireles/crashcrashes

another round attempting to add more polish and get rid of the remaining crashers

AntonioMeireles committed:
fix halt logic

- timeout to cover the case where the VM hanged on shutdown
- only clean cruft if VM really down

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
fix logic issues on VM's boot handling

... and suddently workarounds stop being needed.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
drop workaround for issue i'm unable to reproduce anymore

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
let remain on fs data from problematic VMs

anything that isn't halted (properly) via `coreos halt` is left
under ~/.coreos/running/ as it eases troubleshouting

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #10 from AntonioMeireles/grrr

assorted bug fixing and cleanups

AntonioMeireles committed:
refactor halt logic and get hark kills to actually work

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #11 from AntonioMeireles/fix_halt

refactor halt logic and get hark kills to actually work

AntonioMeireles committed:
grab VM's IP from the VM itself, not from host

en passant clean up cruft and get things a bit simpler

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #12 from AntonioMeireles/getIPfromGuest

grab VM's IP from the VM itself, not from host

AntonioMeireles committed:
add an option to upload file to a VM

per @rimusz request

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #14 from AntonioMeireles/scp

add an option to upload file to a VM

AntonioMeireles committed:
checks that destination dir exists on target before attempting to upload.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #15 from AntonioMeireles/copy_checkdestdirexists

checks that destination dir exists on target before upload attempt

AntonioMeireles committed:
update Godeps'

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
be a bit more flexible when searching running VMs...

... regarding the way `corectl` is invoked

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
increase to 45s the TTL for a VM to boot and we spot its PID and IP

... as not everybody has boxes with fast SSDs

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
refactor UUID handling a bit

also,  drop a sysyscall hack that got commited by mistake

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #16 from AntonioMeireles/goose_chase

yet another round of tweaks for issues spoted by @rimusz

AntonioMeireles committed:
refactor things a bit, no functional changes

- locks, channels, etc are all per VM, now

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #17 from AntonioMeireles/everythingLocal

refactor things a bit, no functional changes

AntonioMeireles committed:
on first run don't break if '/etc/exports' doesn't exist

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #18 from AntonioMeireles/etcExportsFix

on first run don't break if '/etc/exports' doesn't exist

AntonioMeireles committed:
scp: handle case when destination is a directory.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #20 from AntonioMeireles/scpFix

scp: handle case when destination is a directory.

AntonioMeireles committed:
properly signal when we exit on error

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
add an experimental query option

- Work in progress. May or may not fully replace 'ps' in the future

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
Merge pull request #21 from AntonioMeireles/exitWithError

properly signal when we exit on error

AntonioMeireles committed:
Merge pull request #22 from AntonioMeireles/queryFlag

add an experimental query option

AntonioMeireles commented:
@philips, et al. 

TimeOuting. action ongoing at [github.com/TheNewNormal/corectl](github.com/TheNewNormal/corectl)
